Add CI-build script

gunnarx opened the pull request:
Mostly untested, but it starts a build at least.

Proposal for discussion.  Likely needs a couple of updates before merging.

One concern is where the dependent layer are to be placed, but the current expectation seems to be that poky/ meta-openembeded/ and meta-gplv2/ should be adjacent to the meta-ivi "root" (repository working dir):

```
$ cd myworkdir
$ tree -L 2 -d
├── meta-gplv2
│   ├── ... (abbreviated)
├── meta-ivi
│   ├── build
│   ├── meta-ivi
│   ├── meta-ivi-bsp
│   ├── meta-ivi-test
│   └── scripts
├── meta-openembedded
│   ├── ... (abbreviated)
└── poky
│   ├── ... (abbreviated)
```

The script was tested as follows:

```
$ meta-ivi/scripts/ci-build.sh
```
or
```
$ cd meta-ivi
$ scripts/ci-build.sh
```
Both seem to work as they should.

Commit:
Add initial ci-build.sh

This is a copy of ci-build.sh from genivi-dev-platform which is
*not yet adjusted* for meta-ivi build.  This is to facilitate diffing
against the next commit to see what was changed.

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

Commit:
First adjustment of ci-build.sh for meta-ivi

Included new image target for build with sdk, similar to GDP
setup.  The content of that file likely needs some adjustments.

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

gunnarx commented:
I believe it works now.  Please review since there might be some old cruft left from GDP version it was based on.  
- If reading through it makes sense and fix what you can see
- If it builds the image correctly
... then it can be merged

slawr commented:
I need to leave early today but I'll try to kick off a local build for R-Car either today or tomorrow and will let you know the result.

jeremiah commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,512 @@
+
+# This is some trickery, but quite useful.
+# Return the value of a variable whose name is stored in another variable
+deref() { eval echo \$$1 ; }
```
Why not just call ${other_variable}?

gunnarx replied:
${var} returns the value stored in var
deref var returns the value stored in the variable whose name is stored in var
e.g.
```
x=y
y=z
echo $x    <- prints out "y"
echo $(deref x)    <- prints out "z"
```

gunnarx replied:
also
```
x=y
y=z
echo ${$x}  <- you'd think this should work, but...
--> bash: ${$x}: bad substitution
```
But maybe there's another way...

gunnarx replied:
The reason something like it is needed here is that the variable *name* itself is generated by the program (it's not known, like in the above example of x & y).

jeremiah commented on scripts/init.sh:
```diff
@@ -0,0 +1,33 @@
+source poky/oe-init-build-env
+
+echo
+echo "Now run:  bitbake miranda-image"
```
I see this is just a comment but it says "miranda" when other scripts say "pulsar". Does this matter? Will it confuse someone (like it confuses me. :)

gunnarx replied:
This (and all other files) should not be part of PR - only ci-build.sh was intended.
I must have goofed on the last commit.  I'll fix and update.

gunnarx commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,512 @@
+  local names="$1" name clean_name suffix var_name value d
+
+  for name in $names ; do
+    clean_name=$(echo "$name" | sed 's/-/_/g')
```
Could more succinctly use string substitution directly in bash, it's just that I'm familiar with the sed syntax...

gunnarx commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,512 @@
+
+  # Variables cannot have dashes in name therefore, underscores must be
+  # used.  For example: LAYER_meta_browser_COMMIT for meta-browser layer
+  clean_name=$(echo "$name" | sed 's/-/_/g')
```
Could use string substitution directly in bash, it's just that I'm familiar with the sed syntax...

gunnarx replied:
... patches welcome :)

Commit:
checkout script: remove bash specific syntax

Script used >& syntax but did not specify a shell that is guaranteed
to understand it (csh, bash, zsh).  This triggers the message
"Syntax error: Bad fd number" on some build machines.

An alternative would be to use /bin/bash explicitly since ci-build
requires bash anyway, but maybe checkout might be used on its own.
This is just as easy.

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

gunnarx commented:
> This (and all other files) should not be part of PR - only ci-build.sh was intended. I must have goofed on the last commit. I'll fix and update.

The image definition files are now gone.

gunnarx commented:
Build script seems to work now, according to [this test pipeline](https://go.genivi.org/go/tab/build/detail/meta-ivi-with-script/13/Build/1/init_and_bitbake).  (A few comments were removed after that, but it should be the same script)

Might be some corner-case bugs remaining of course. 

Some instructions to copy artifacts to a staging/ directory should be added but I'll let @frznlogic consider how to add that later.  

I recommend merging now.

frznlogic commented:
Hi, just saw that you mentioned it should be pulled in now.

Some general comments/questions. 
Wouldn't it be better to break the scripts into their own repository for code reusability and to avoid duplication of efforts? Ie, a separate repository containing this script modified to be more general,  provide configurations within the repo or the go.cd environment for what to fetch etc. 

The bash code is a bit under-documented, some comments on the intention of the functions would be nice. Please explain the trickery and the rationale instead of calling it trickery also :)

frznlogic commented:
Also, please remove the WIP in the PR name :)

gunnarx commented:
> Wouldn't it be better to break the scripts into their own repository for code reusability and to avoid duplication of efforts? 

You mean the script used by meta-ivi and GDP?  Yes, there is a bit of duplication between them, but I felt the argument to keep them with the source (below) stronger so far.  I am also unsure how different GDP & meta-ivi scripts will become, or continue to be.  (Related to the strategy of parent-project for meta-ivi, with submodules, etc.) so this is a start, with potential for refactoring later.

> Ie, a separate repository containing this script modified to be more general, 

Basically my argument against this has been that ci-build.sh *can* be usable also for advanced users (and I hope we can take any steps to improve that usability, such as documentation).   For me that points towards keeping the build-script with the source that users already have, instead of requiring users to pick up another new repository.  We also used to have that, a "baseline-ci-helper" repo for Go builds, but it was just overhead...

There are a few dependencies between script & source also (for example the list of sub-layers inside the script).  When there's a dependency that needs to be in sync with eachother, it points towards keeping it in one repo if possible.  There's the possibility of a "scripts" submodule perhaps, but I don't know if we are there yet.

> provide configurations within the repo or the go.cd environment for what to fetch etc.

I find I often prefer a straight-forward approach, using git branches and such, to manage differences, instead of creating too much structure.  Unless the problem/project is big enough to actually gain speed in the long run by creating it.  For this you need to decide on some "configuration language" of sorts, then keep those configurations in sync with each project, and so on.
I'm not sure this build stuff big enough  to warrant the effort.  BUT, if you envision something larger here, like reuse in pelux as well then I might be convinced I guess. I'm open to suggestions.

> The bash code is a bit under-documented, some comments on the intention of the functions would be nice. 

Agreed, but could it be future improvement?  (I was hoping you take over the script, assuming it gets merged here...).   If you review-comment on the lines that you feel must be fixed, I'll address those.
**EDIT** Added comments in bf5e8d3

> Please explain the trickery and the rationale instead of calling it trickery also :)

In my opinion, the comment just above the deref function explains what the function does as clearly as can be done.  My answer to Jeremiah basically repeated the comment. The need for the function can be seen where it is used, but OK, I'll add rationale.

Commit:
ci-build: Documenting the script

gunnarx commented:
@frznlogic Please review latest commit for more documentation.  We could squash before merge if you prefer.

slawr commented:
>>     Ie, a separate repository containing this script modified to be more general,
> 
> Basically my argument against this has been that ci-build.sh can be usable also for advanced users (and I hope we can take any steps to improve that usability, such as documentation). For me that points towards keeping the build-script with the source that users already have, instead of requiring users to pick up another new repository.

If its simple to use I could see that having it in the source may help contribution, including component maintainers, if it simplifies setup of the baseline when they are not used to doing it.

gunnarx commented:
Well, yes the script is simple(r) to use than the alternatives.  Currently the script checks out the right commits (taken from README), but does *not* clone the layers themselves.  I think that should be added, unless we have some other parent-project setup.

So as of now: these are basically the minimal steps you must do:
```bash
$ git clone git://github.com/GENIVI/meta-ivi
$ git clone git://git.yoctoproject.org/poky
$ git clone git://git.openembedded.org/meta-openembedded
$ git clone git://git.yoctoproject.org/meta-gplv2
$ export TARGET=qemuarm64
$ meta-ivi/scripts/ci-build.sh
```
That should start the build.  It still takes some steps away.  Steps 2-4 ought to be eliminated in some way - please let me know how you think it should be done.

slawr commented:
Thanks for the note and sorry for not trying the script earlier like I said I would. It slipped my mind with all the Domain Interaction work recently. The note makes it clear where I would do the extras for a community board.

When I think of baseline builds I see two broad use cases, building the "as is" baseline and the baseline as part of something larger. The former could be for the official qemu or community targets. Community targets might mean the addition of a layer or two but you are not radically reshaping the distro. For me they are the principal targets of such "helpers". It may simply be that the helper needs to be flexible enough to support the community targets either by extension or manual local work to add things.

Personally I would not worry too much about the second use case, baseline as part of something much larger, currently. Such a distro is likely to have its own tooling anyway.

frznlogic commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+define_with_default MIRROR "https://docs.projects.genivi.org/releases/yocto_mirror"
+define_with_default PREMIRROR ""  # By default none (but we have the shared DL_DIR)
+define_with_default RM_WORK false
+define_with_default REUSE_STANDARD_DL_DIR true
```
I'd prefer the default for DL_DIR to be false

gunnarx replied:
OK

frznlogic commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+define_with_default PREMIRROR ""  # By default none (but we have the shared DL_DIR)
+define_with_default RM_WORK false
+define_with_default REUSE_STANDARD_DL_DIR true
+define_with_default REUSE_STANDARD_SSTATE_DIR true
```
I'd prefer the default for SSTATE_DIR to be false. 

It's easy to set this for the CI build and generates a fairly complex error message for whatever user right now.

gunnarx replied:
OK

frznlogic commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+AGENT_STANDARD_DL_DIR="/var/cache/yocto/downloads"
+AGENT_STANDARD_SSTATE_DIR="/var/cache/yocto/sstate"
+AGENT_STANDARD_SGX_LOCATION="/var/go/sgx_bin"
+AGENT_STANDARD_SGX_GEN3_LOCATION="/var/go/sgx_bin_gen3"
```
All of these paths are specific to the CI environment on go.genivi.org. I'd prefer them to be either be tied to $HOME or $PWD if possible and then change them locally in the CI environment.

gunnarx replied:
These have the prefix AGENT_ because they are for the build agents.  They affect the user build only depending on REUSE_*, and if the user does not set its own SGX_DRIVERS location, etc.
So I propose to keep them.  We could set a user-friendly default, but what would that be?
Proposal?  $HOME/something?

frznlogic commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+    echo -n "Appending to local.conf: "
+    cat <<EOT | tee -a "$LOCAL_CONF"
+$1
+EOT
```
The script will continue appending to an existing local.conf file for every time you try to run it. Ie, we wind up with the MIRRORS_append once every time we try to run this script as it is now.

gunnarx replied:
Agreed, this is because CI environments are thrown away frequently.  So this could be improved for end-user, yes.

frznlogic commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+
+D=$(dirname "$0")
+cd "$D"
+BASEDIR=$(readlink -f "$PWD/..")
```
For some reason we wind up building in <path>/meta-ivi/build/... I'd prefer to keep the build directory in <path>/build, the same level as the other meta-layers.

gunnarx replied:
OK, sure, that could be changed.  This is a remnant from GDP where build/ and scripts/ are in parallel (since GDP git is a parent project but meta-ivi is just a layer).

frznlogic reviewed (changes requested):
Looks better with the comments etc. I've tried running the script and had some more minor issues with the defaults basically. What it boils down to imho, it would be nicer to move the burden of handling defaults from the users to the CI environment.

gunnarx commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+set +e  # Allow failures temporarily
+rm -rf staging
+shopt -s nullglob
+stage_artifact mv build/tmp/deploy/licenses
```
By the way, there isn't very deep thought behind the selection of move vs copy here.  Mostly I guess the idea should be to move large things instead of copy, to save disk space and time but to copy things like logs that are expected to be found when poking around the build tree...

All of this is sort of optimized for CI and we might want to put a condition on it.

gunnarx commented on scripts/ci-build.sh:
```diff
@@ -0,0 +1,562 @@
+define_with_default REUSE_STANDARD_DL_DIR true
+define_with_default REUSE_STANDARD_SSTATE_DIR true
+define_with_default SGX_DRIVERS $AGENT_STANDARD_SGX_LOCATION
+define_with_default SGX_GEN_3_DRIVERS $AGENT_STANDARD_SGX_GEN3_LOCATION
```
Note that once the default for "Use standard dirs" is false, the user will be able to specify their own values for SGX bin, DL_DIR an SSTATE_DIR.

define_with_default() will not change a variable if it already has a value in environment.

Commit:
ci-build: Change default for reused (fixed) directories to false

There are some "fixed" locations for caching (downloads, sstate) used for
all builds by build agents.  These used to be used by default, and could
be overridden to not be used.  The logic is now reveresed to be better
optimized for end-user usage, and less optimized for CI.

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

Commit:
ci-build: Change build/ dir location to parent dir

The script used to place build inside of meta-ivi, (a remnant
from GDP where the script is in a "parent" container project).
The right location ought to be the directory just above meta-ivi
layer directory.

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

Commit:
ci-build: Avoid multiple appends to conf files

Signed-off-by: Gunnar Andersson <gandersson@genivi.org>

gunnarx commented:
Squashed one commit that was superfluous.  Question now is if we squash more or keep this logical history.
@frznlogic I'm not sure if your review is still valid or can be approved now?  I think you need to try the script and decide if it's OK.

frznlogic commented:
@gunnarx I'm good with it, let's get this in imho. Waiting for the builds to finish and then pulling it in :)
